Preparing optwin.vim for subsequent translations.

RestorerZ opened the pull request:
The main idea of this pull request is to separate code and text data that require translation.

All the code (well, almost all of it) is in the optwin.vim file, and the text data is in the translation file located in the /lang directory.

When you create a options window for the command `:options`, code optwin.vim file searches in the directory /lang file translation, and if found, comparing with the actual English version that shows the translated content options window.
If any of these conditions are not met, i.e. the translation file is missing or it does not match the English version, the current version in English is displayed.
For English-speaking users, of course, the English version of the options window will be displayed.

Основная идея этого запроса на принятие в разделение кода и текстовых данных, требующих перевода.

Весь код (ну, почти весь) находится в файле optwin.vim, а текстовые данные в файле с переводом, расположенном в каталоге /lang.

При создании окна параметров по команде `:options`, код файла optwin.vim ищет в каталоге /lang файл с переводом и, если находит, сравнив с актуальной английской версией, показывает переведённое содержание окна параметров.
При невыполнении любого из этих условий, то есть отсутствие файла перевода или его несоответствие английской версии, отображается актуальная версия на английском языке.
Для англоговорящих пользователей будет, естественно, отображаться английский вариант окна параметров.

File for translation. English version.
The file contains text data displayed in the options window.

RestorerZ committed:
File for translation. English version.

File for translation. English version.
The file contains text data displayed in the options window.

Файл для перевода. Английская версия.
Файл содержит текстовые данные, отображаемые в окне параметров.

RestorerZ committed:
Merge pull request #1 from RestorerZ/RestorerZ-patch-1

File for translation. English version.

RestorerZ committed:
Preparing optwin.vim for subsequent translations.

Preparing optwin.vim for translations.

The main idea of this pull request is to separate code and text data that require translation.

All the code (well, almost all of it) is in the optwin.vim file, and the text data is in the translation file located in the /lang directory.

When you create a options window for the command `:options`, code optwin.vim file searches in the directory /lang file translation, and if found, comparing with the actual English version that shows the translated content options window.
If any of these conditions are not met, i.e. the translation file is missing or it does not match the English version, the current version in English is displayed.
For English-speaking users, of course, the English version of the options window will be displayed.

Подготовка optwin.vim для последующих переводов.
Подготовка optwin.vim для переводов.

Основная идея этого запроса на принятие в разделение кода и текстовых данных, требующих перевода.

Весь код (ну, почти весь) находится в файле optwin.vim, а текстовые данные в файле с переводом, расположенном в каталоге /lang.

При создании окна параметров по команде `:options`, код файла optwin.vim ищет в каталоге /lang файл с переводом и, если находит, сравнив с актуальной английской версией, показывает переведённое содержание окна параметров.
При невыполнении любого из этих условий, то есть отсутствие файла перевода или его несоответствие английской версии, отображается актуальная версия на английском языке.
Для англоговорящих пользователей будет, естественно, отображаться английский вариант окна параметров.

RestorerZ committed:
Merge pull request #2 from RestorerZ/RestorerZ-patch-2

Preparing optwin.vim for subsequent translations.

brammool commented:
I'm afraid I don't like this much.  I would very much like to:
- Keep the English text in one place, possibly add markers like we have N_() in C source code.
- Be able to extract the English text, like we do with C source code
- have "make xx" in src/po merge these text messages with the others
- translate the xx.po file as usual.
- use a builtin function to get the translation (it's in the todo file somewhere).

This may be more work up front, but will make updating translations many times easier.

RestorerZ committed:
Preparing optwin.vim for subsequent translations.

A line was omitted when editing
call append("$", " \tset debug=" . &debug)

RestorerZ committed:
Merge pull request #3 from RestorerZ/RestorerZ-patch-3

Preparing optwin.vim for subsequent translations.

RestorerZ commented:
Yes, I fully agree on updating translations. The variant I suggested is very inconvenient in this.
Initially, I thought to use the menutranslate command for translation.
I'll see what built-in function you're talking about.

Да, по обновлению переводов полностью согласен. Предложенный мной вариант очень неудобен в этом.
Первоначально я думал использовать для перевода команду menutranslate.
Я посмотрю о какой встроенной функции вы говорите.

brammool commented:
Please have a look at patch 8.2.1544.  I tried it out with Dutch and it works for me.

One disadvantage is that in the current form line breaks can't be moved around.  That is a bit of a problem when the message gets longer after translation.

RestorerZ commented:
Bram, thank you for the gettext()!

And then, not finding anything in VIM for translation, I sat down to study
the VimL tried to write something taking into account what you said.
And even something [turned out](http://www.mediafire.com/file/ha1wqz76eoj70hk/file). :)
To be honest, I'm not a real programmer, and this is probably the third time
I've encountered VimL.  Therefore, but not everything is finished.
And today I was going to update the VIM and then this news!
Thank you Bram again for the wonderful editor and what you do!

Bram, you will have an error at this point if you use a non-English text:
```
fun! <SID>Find(lnum)
    if getline(a:lnum - 1) =~ "(local to"
      let local = 1
```
When I wrote the script for translating, I did this:
```
fun! <SID>Find(lnum)
   if <SID>ChckOptIsLocl(a:lnum)
    let local = 1
```   
```
fun! <SID>ChckOptIsLocl(lnum)
  let lnum = a:lnum
  let line = getline(lnum-1)
  while match(line, "^[a-z]")
    let lnum = lnum - 1
    let line = getline(lnum)
  endwhile
  let name = substitute(line, '\([^\t]*\).*', '\1', "")
  if (index(s:isloclopt, name)) >= 0 
    return 1
  endif
    return 0
endfun
```
```
function! <SID>PrtOptDesc(optname)
  if match(a:optname, "^L-") >= 0
    let optname = trim(a:optname, "L-", 1)
    call add(s:isloclopt, optname)
  else
    let optname = a:optname
  endif
```
```
call <SID>Header(2)
call <SID>PrtOptDesc("L-whichwrap")
call <SID>OptionL("ww")
```
It's probably not very optimal, but I couldn't think of any other way.

P. S.
The documentation [says](https://vimhelp.org/options.txt.html#'whichwrap') that the option 'whichwrap' has a global scope, and in
the option window, that it has a local scope for window.  Who to trust?


Брам, спасибо за gettext()!

А то, не найдя в VIM ничего для перевода, я сел за изучение VimL и попытался
что-то написать с учётом того, что вы говорили.
И  даже что-то [получилось](http://www.mediafire.com/file/ha1wqz76eoj70hk/file). :)
Честно говоря, я не настоящий программист, да и с VimL это, наверное, раз
третий, когда я сталкиваюсь. Поэтому не всё ещё доделано.
А сегодня решил обновить VIM и тут такая новость!
Спасибо Брам ещё раз за чудесный редактор и то, что вы делаете!

Брам, у вас в этом месте будет ошибка, если использовать не английский текст:
```
fun! <SID>Find(lnum)
    if getline(a:lnum - 1) =~ "(local to"
      let local = 1
```
Я когда писал скрипт для перевода, то сделал так:
```
fun! <SID>Find(lnum)
   if <SID>ChckOptIsLocl(a:lnum)
    let local = 1
```    
```
fun! <SID>ChckOptIsLocl(lnum)
  let lnum = a:lnum
  let line = getline(lnum-1)
  while match(line, "^[a-z]")
    let lnum = lnum - 1
    let line = getline(lnum)
  endwhile
  let name = substitute(line, '\([^\t]*\).*', '\1', "")
  if (index(s:isloclopt, name)) >= 0 
    return 1
  endif
    return 0
endfun
```
```
function! <SID>PrtOptDesc(optname)
  if match(a:optname, "^L-") >= 0
    let optname = trim(a:optname, "L-", 1)
    call add(s:isloclopt, optname)
  else
    let optname = a:optname
  endif
```
```
call <SID>Header(2)
call <SID>PrtOptDesc("L-whichwrap")
call <SID>OptionL("ww")
```

Наверное, это не очень оптимально, но по-другому я не придумал.

P. S.
В документации [написано](https://vimhelp.org/options.txt.html#'whichwrap'), что параметр 'whichwrap' имеет глобальную область
действия, а в окне параметров, что локальную для окна.  Кому верить?

brammool commented:
I found a simpler way: First get the translated message for "local to" and use them in many places.
This also avoids having to call gettext() many times.

Commit:
Merge remote-tracking branch 'upstream/master' into master

Commit:
Preparing optwin.vim for subsequent translations.

Placed a call to the gettext() function for text that requires translation.
Put a line with the output of the 'termwinscroll' parameter value in the place where
it should be.

Расставил вызов функции gettext() для текста, который требует перевода.
Поставил строку с выводом значения параметра 'termwinscroll' на то место, где
она должна быть.

Commit:
Removed optwin_en.latin1.vim

No more need to use this file.

Удалил optwin_en.latin1.vim

Больше не требуется этот файл.

RestorerZ commented:
Bram, yes, the string length is really not that simple. It is not always
possible to fit the translation into the specified length without losing the
meaning of the phrase.

The first thing that comes to mind is to apply something like a formatting
operator `gqgq` to the inserted string. But this will probably not always be done
correctly.  Alternatively, the translator places the end-of-line character "\n"
according to the required length of the string, and then VIM will go through the
string and cut it according to these labels. Well, something like that:
```
while strlen(line) > &tw
    let subline = matchstr(line, '^[ \t]\w{-1,}\n')
    let line = trim(line, subline, 1)
    let subline = trim(sublie, '\n', 2)
    call add(printlns, subline)
endwhile
append("$", printlns)
```
However, this is only required when the function `append()`outputs a string.
The `:echo` command splits the string itself if it is enclosed in double quotes.
Isn't that right? Also, the formatting of the entire printed text may be lost if
the translation line is too long.  In general, you are right, not everything is
so simple.
Your solution for checking local functions is great.

Брам, да, с длиной строки действительно не всё так просто. Не всегда получается
уместить перевод в заданную длину не потеряв при этом смысл фразы.

Первое, что приходит на ум — это применить к вставленной строке что-то типа
оператора форматирования `gqgq`.  Но это, скорее всего, будет не всегда корректно
сделано.
Ещё, как вариант, переводчик расставляет символ конца строки "\n" в соответствии
с требуемой длиной строки, а потом VIM пройдётся по строке и разрежет её по этим
меткам. Ну как-то так:
```
while strlen(line) > &tw
    let subline = matchstr(line, '^[ \t]\w{-1,}\n')
    let line = trim(line, subline, 1)
    let subline = trim(sublie, '\n', 2)
    call add(printlns, subline)
endwhile
append("$", printlns)
```
Но это требуется только при выводе строки функцией `append()`.
Команда `:echo` сама разбивает строку, если она заключена в двойные кавычки. Ведь
так, да? И ещё может сбиться форматирование всего печатаемого текста, если
строка перевода будет слишком длинной. В общем, вы правы, не всё так просто.
Ваше решение по проверке локальных функций великолепное.

brammool commented:
What we should do is, instead of using append() directly, define a function that takes the name of the option and the explanation as two arguments.  The explanation should be one text string, so that it can be translated as one entry.  Then the translator can put "\n" line break characters as needed.  The function can take the (translated) text, split at the line breaks, and insert the text with additional indent.  The existing explanations that are already two lines need to be changed to one text string.

brammool commented:
Patch 8.2.1652 makes this work.  It has three example options already done.

Commit:
Merge remote-tracking branch 'remotes/upstream/master' into master

# Conflicts:
#	runtime/optwin.vim

Commit:
Preparing optwin.vim for subsequent translations

Placed a call function AddOption() instead of the function append() for options
with their brief description

Расставил вызов функции AddOption() вместо функции append() для опций с их
кратким описанием
